### DISCUSSION TITLE
locked database SQLITE3

### DISCUSSION THREAD

**Discussion by fedeball** [sqlitebrowser/sqlitebrowser#3225] [Q&A]
HI,
I have a database that get locked every time I use it.
It works perfectly the first time it is created, modified and populated but the second time I get in it is locked.
I read that is because two process likes to get into the DB at the same time but my program has no other process that the one I'm using. I verified all and all processes are closed before exiting them.

Questions are, 
Is there any other explanation for locking DB that the two proccesses explanation?
how can I see when the DB is blocked??
How can I unlock it ?
Is there any way to know when the DB get locked???

Thanks a lot for your help
Fede

**Comment by chrisjlocke** [comment]
You mention a database, then 'my program'.  How do you know the database is getting locked?  What are you doing, what did you expect, and what actually occurred?
In DB Browser for SQLite, when you write to the database, you have to 'commit' it.  

![image](https://user-images.githubusercontent.com/7872218/202445393-6562cd6b-d5d0-492d-8e2d-5b6161b32fd8.png)

In this instance, 'write changes' appears as a button at the top of the screen.  Click this, and the database is unlocked again.

**Reply by fedeball** [reply]
Hi and thanks for your help.
I'm actually doing a c++ program that Create, populate a DB in order to add new columns.
It was working well until I tried to ALTER TABLE adding a new column.
Whenever I do this, the DB get locked.
I went to de DB Browser for SQLITE but the Write Change is disable.
below the process that lock the DB.

void write_MPed_sector(int fecha, string exchange, string sector, string info, float dato){

    sqlite3* db;
    int exit=0;
    string direccion=DIRECCION_Datos_Pref_SECTOR;
    direccion =direccion + exchange.c_str()+ ".db";
    char* char_arr;
    char_arr = &direccion[0];
    exit=sqlite3_open(char_arr,&db);
    char *messaggeError;
    
    string sql="CREATE TABLE IF NOT EXISTS "+sector+" (FECHA INTEGER PRIMARY KEY);";
    exit=sqlite3_exec(db,sql.c_str(),NULL,0,&messaggeError);
    sql=" ALTER TABLE "+sector+" ADD COLUMN "+info+" VARCHAR;";
    exit=sqlite3_exec(db,sql.c_str(),NULL,0,&messaggeError);
    sql="INSERT OR IGNORE INTO "+sector+" (Fecha,"+info+") VALUES ('"+to_string(fecha)+"','"+to_string(dato)+"');";
    exit=sqlite3_exec(db,sql.c_str(),NULL,0,&messaggeError);
    sql="UPDATE "+sector+" SET "+info+" = '"+to_string(dato)+"' WHERE Fecha = "+to_string(fecha)+";";
    exit=sqlite3_exec(db,sql.c_str(),NULL,0,&messaggeError);

    sqlite3_close(db);
}

**Reply by chrisjlocke** [reply]
I don't know C++ I'm afraid, but the guys on the SQLite forum might be able to help.  They have sample programs and might spot where you're going wrong.  
https://sqlite.org/forum/forummain

Mentioning DB Browser for SQLite is muddying the waters though - this has nothing to do with DB4S if this is your program locking your database - I assume if you close DB4S and run your program, you still get the same issue?

Again, ignoring your program, you should be able to modify the table, add a column, and click OK.  You don't get a 'write changes' button as the changes are implied by clicking OK.  The same if you rename a column.  Creating a new table however, does require you to click 'Write changes'.  No, I don't know either...
